prow/gerrit/gerrit.go

```diff
@@ -0,0 +1,189 @@
+	syncTime := time.Now()
+	changes, err := c.QueryChanges(rateLimit)
+	if err != nil {
+		return fmt.Errorf("failed query changes : %v", err)
+	}
+
+	for _, change := range changes {
+		if err := c.ProcessChange(change); err != nil {
+			c.log.WithError(err).Error("Failed process change %v", change.CurrentRevision)
+		}
+	}
+
+	c.lastUpdate = syncTime
+	c.log.Infof("Processed %d changes", len(changes))
+	return nil
+}
+
+// QueryChanges will query all gerrit changes since controller's last sync loop
+func (c *Controller) QueryChanges(rateLimit int) (map[string]gerrit.ChangeInfo, error) {
+	opt := &gerrit.QueryChangeOptions{}
+	opt.Query = []string{}
+	for _, proj := range c.projects {
+		opt.Query = append(opt.Query, "project:"+proj)
+	}
+	opt.AdditionalFields = []string{"CURRENT_REVISION"}
+
+	done := false
+
+	// store a map of changeID:change
+	pending := map[string]gerrit.ChangeInfo{}
+	start := 0
+
+	for !done {
+		opt.Limit = rateLimit
+		opt.Start = start
+
+		// The change output is sorted by the last update time, most recently updated to oldest updated.
+		// Gerrit API docs: https://gerrit-review.googlesource.com/Documentation/rest-api-changes.html#list-changes
+		changes, _, err := c.gc.QueryChanges(opt)
+		if err != nil {
+			// should not happen? Let next sync loop catch up
+			return pending, fmt.Errorf("failed to query gerrit changes: %v", err)
+		}
+
+		if len(*changes) == 0 {
+			break
+		}
+
+		for _, change := range *changes {
+			// if we already processed this change, then we stop the current sync loop
+			layout := "2006-01-02 15:04:05"
+			updated, err := time.Parse(layout, change.Updated)
+			if err != nil {
+				c.log.WithError(err).Error("Parse time %v failed", change.Updated)
+				start++
+				continue
+			}
+
+			// process if updated later than last updated
+			// stop if already parsed
+			if updated.After(c.lastUpdate) {
+				// in case of a new change comes in during the current sync cycle
+				if _, ok := pending[change.CurrentRevision]; !ok {
+					pending[change.CurrentRevision] = change
+				}
+				start++
+			} else {
+				done = true
+				break
+			}
+		}
+	}
+
+	return pending, nil
+}
+
+// ProcessChange creates new presubmit prowjobs base off the gerrit changes
+func (c *Controller) ProcessChange(change gerrit.ChangeInfo) error {
+	rev, ok := change.Revisions[change.CurrentRevision]
+	if !ok {
+		return fmt.Errorf("cannot find current revision for change %v", change.ID)
+	}
+
+	for _, spec := range c.ca.Config().Presubmits[change.Project] {
+		kr := kube.Refs{
+			Org:     c.instance,
+			Repo:    change.Project,
+			BaseRef: rev.Ref,
+			BaseSHA: change.CurrentRevision,
+			Pulls: []kube.Pull{
+				{
+					Number: change.Number,
+					Author: change.Owner.Name,
+					SHA:    rev.Ref,
+				},
+			},
+		}
+		pj := pjutil.NewProwJob(pjutil.PresubmitSpec(spec, kr), map[string]string{})
+		c.log.WithFields(pjutil.ProwJobFields(&pj)).Info("Creating a new prowjob.")
```

Comment:
It would probably be helpful to include the change ID as a field in this log message and any other log messages that are specific to a change. That will allow us to query for the logs for actions and errors associated with a specific change.